### PR TITLE
Refactored core-data clients

### PR DESCRIPTION
**Pull request by jduranf**
This is the first part of #609.
If everything it's fine, I will refactor the other micro service clients in the same way.

Signed-off-by: Joan Duran <jduran@circutor.com>

### CODE REVIEW

File: pkg/clients/coredata/event.go [hunk 1]
```diff
@@ -72,426 +68,86 @@ func (e *EventRestClient) init(params types.EndpointParams) {
...
-		if err != nil {
-			fmt.Println(err.Error())
-			return models.Event{}, err
-		}
-
-		return models.Event{}, types.NewErrServiceClient(resp.StatusCode, bodyBytes)
-	}
-
-	return e.decodeEvent(resp)
+	return e.requestEvent(e.url + "/" + id)
```
Comment (drasko) [inline pkg/clients/coredata/event.go]: I do not like this sting constant hardcoding. Also, I find `fmt.Sprintf()` more readable here.

File: pkg/clients/coredata/event.go [hunk 2]
```diff
@@ -72,426 +68,86 @@ func (e *EventRestClient) init(params types.EndpointParams) {
...
-
-	if resp.StatusCode != http.StatusOK {
-		return 0, types.NewErrServiceClient(resp.StatusCode, bodyBytes)
-	}
-	count, err := strconv.Atoi(bodyString)
-	if err != nil {
-		return 0, err
-	}
-	return count, nil
+	return countRequest(e.url + "/count")
```
Comment (drasko) [inline pkg/clients/coredata/event.go]: Here also - hardcoded string constant.

File: pkg/clients/coredata/event.go [hunk 3]
```diff
@@ -72,426 +68,86 @@ func (e *EventRestClient) init(params types.EndpointParams) {
...
-	if resp.StatusCode != http.StatusOK {
-		return 0, types.NewErrServiceClient(resp.StatusCode, bodyBytes)
-	}
-
-	count, err := strconv.Atoi(bodyString)
-	if err != nil {
-		return 0, err
-	}
-	return count, nil
+	return countRequest(e.url + "/count/" + url.QueryEscape(deviceId))
```
Comment (drasko) [inline pkg/clients/coredata/event.go]: And here.

File: pkg/clients/coredata/event.go [hunk 4]
```diff
@@ -72,426 +68,86 @@ func (e *EventRestClient) init(params types.EndpointParams) {
...
-		bodyBytes, err := getBody(resp)
-		if err != nil {
-			fmt.Println(err.Error())
-			return []models.Event{}, err
-		}
-
-		return []models.Event{}, types.NewErrServiceClient(resp.StatusCode, bodyBytes)
-	}
-	return e.decodeEventSlice(resp)
+	return e.requestEventSlice(e.url + "/device/" + url.QueryEscape(deviceId) + "/" + strconv.Itoa(limit))
```
Comment (drasko) [inline pkg/clients/coredata/event.go]: And here is nice example where `fmt.Sprintf()` would be really beneficial.

File: pkg/clients/coredata/event.go [hunk 5]
```diff
@@ -72,426 +68,86 @@ func (e *EventRestClient) init(params types.EndpointParams) {
...
-		bodyBytes, err := getBody(resp)
-		if err != nil {
-			fmt.Println(err.Error())
-			return []models.Event{}, err
-		}
-
-		return []models.Event{}, types.NewErrServiceClient(resp.StatusCode, bodyBytes)
-	}
-	return e.decodeEventSlice(resp)
+	return e.requestEventSlice(e.url + "/" + strconv.Itoa(start) + "/" + strconv.Itoa(end) + "/" + strconv.Itoa(limit))
```
Comment (drasko) [inline pkg/clients/coredata/event.go]: Here also.

File: pkg/clients/coredata/event.go [hunk 6]
```diff
@@ -72,426 +68,86 @@ func (e *EventRestClient) init(params types.EndpointParams) {
...
-		bodyBytes, err := getBody(resp)
-		if err != nil {
-			fmt.Println(err.Error())
-			return []models.Event{}, err
-		}
-
-		return []models.Event{}, types.NewErrServiceClient(resp.StatusCode, bodyBytes)
-	}
-	return e.decodeEventSlice(resp)
+	return e.requestEventSlice(e.url + "/device/" + url.QueryEscape(deviceId) + "/valuedescriptor/" + url.QueryEscape(vd) + "/" + strconv.Itoa(limit))
```
Comment (drasko) [inline pkg/clients/coredata/event.go]: Again, hardcoded string constants :).

**Comment by jduranf** [comment]
@drasko although I agree with you that using hard-coded string constants should be avoided, the intention of this PR is not fix this issue. The hard-coded string constants where present before this PR, so I disagree that this is a good reason to reject this PR.

Anyway, if we want to define some strings like you suggested in this line:
```
return e.requestEventSlice(e.url + "/device/" + url.QueryEscape(deviceId) + "/valuedescriptor/" + url.QueryEscape(vd) + "/" + strconv.Itoa(limit))
```
We should also use the same defined strings in internal/core/data/router.go:
```
e.HandleFunc("/device/{deviceId}/valuedescriptor/{valueDescriptor}/{limit:[0-9]+}", readingByDeviceFilteredValueDescriptor).Methods(http.MethodGet)
```
But as I said, this is out of the scope of this PR.

File: pkg/clients/coredata/request.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
+/*******************************************************************************
+ * Copyright 2017 Dell Inc.
```
Comment (tsconn23) [inline pkg/clients/coredata/request.go]: @jpwhitemn How should the copyright work here? This is a new file created by @jduranf but AFAIK he's working on the project himself and his employer isn't a member company. Does that mean he should assign copyright to himself? If his employer IS a member company then their name should be here, correct?

Comment (jpwhitemn) [reply pkg/clients/coredata/request.go]: Yes - Joan, just add your own copyright below the Dell one.  See https://github.com/edgexfoundry/device-sdk-go/blob/master/config.go as an example.

Something like this:

// -*- mode: Go; indent-tabs-mode: t -*-
//
// Copyright (C) 2017 Dell Inc.
// Copyright (C) 2018 Joan Duran
//
// SPDX-License-Identifier: Apache-2.0

**Comment by jduranf** [comment]
@tsconn23 fixed. Also fixed some golint issues (one day we should fix all golint issues...).

**Pull request merged**